### PR TITLE
Modify registry scan algorithm

### PR DESCRIPTION
**Pull request by Molter73**
|Related issue|
|---|
|#5864|

## Description

The purpouse of this PR is to modify the scan algorithm in order for it to comply with the new requirements established in issued #5576.

This PR depends heaviliy on the PRs associated with issues #5867 and #5865. Due to this dependency, this PR won't compile on its own, since it has several place holders that must be filled with changes coming from those issues.

The following features are added by this PR:
- Rewrote the scan algorithm for Windows registry almost entirely.
- Keys and values are treated as separate entities, allowing for alerts to be generated on them individually.
- Added separate logic for registry event generation.
- Added `ignore`, `recursion_level` and `restrict` checks on the scan algorithm, once those values are gathered from the configuration they should work.

### CODE REVIEW

File: src/syscheckd/registry/registry.c [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+    }
+
+    if (value_count) {
+        fim_read_values(current_key_handle, &new, &saved, arch, value_count, mode);
+    }
+
+    fim_registry_free_key(new.registry_entry.key);
+    fim_registry_free_key(saved.registry_entry.key);
+    RegCloseKey(current_key_handle);
+    return;
```
Comment (jesusjimsa) [inline src/syscheckd/registry/registry.c:457]: There is no need for an empty return statement at the end of a void function, please consider removing it.

**Review by jesusjimsa — CHANGES_REQUESTED** [review]
Only two of the new functions have _doxygen_ information, please add this information to the rest of them.

**Review by antoniomanuelfr — APPROVED** [review]
LGTM

File: src/syscheckd/registry/events.c [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+                                     fim_event_mode mode,
+                                     unsigned int type,
+                                     __attribute__((unused)) whodata_evt *w_evt,
+                                     const char *diff) {
+    cJSON *changed_attributes;
+    char path[OS_SIZE_512];
+
+    if (old_data != NULL) {
+        changed_attributes =
+        fim_registry_compare_value_attrs(new_data->registry_entry.value, old_data->registry_entry.value, configuration);
```
Comment (jesusjimsa) [inline src/syscheckd/registry/events.c:90]: Please, move the `fim_registry_compare_value_attrs` function to the same line as the `changed_attributes` variable.
```suggestion
        changed_attributes = fim_registry_compare_value_attrs(new_data->registry_entry.value,
                                                              old_data->registry_entry.value, configuration);
```

File: src/syscheckd/registry/registry.c [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+    if (fim_registry_validate_path(value_path, configuration)) {
+        return;
+    }
+
+    if (fim_check_restrict(value_path, configuration->filerestrict)) {
+        return;
+    }
+
+    saved->registry_entry.value =
+    fim_db_get_registry_data(syscheck.database, new->registry_entry.key->id, new->registry_entry.value->name);
```
Comment (jesusjimsa) [inline src/syscheckd/registry/registry.c:285]: Please, move the function to the line above.
```suggestion
    saved->registry_entry.value = fim_db_get_registry_data(syscheck.database, new->registry_entry.key->id,
                                                           new->registry_entry.value->name);
```

**Review by jesusjimsa — CHANGES_REQUESTED** [review]

**Review by jesusjimsa — CHANGES_REQUESTED** [review]
Please, consider moving the _debug_, _error_ and _warning_ messages to the files in the _error_messages_ folder in order to not have them hardcoded.

File: src/syscheckd/create_db.c [hunk 4]
```diff
@@ -98,6 +98,10 @@ void fim_scan() {
         nodes_count = fim_db_get_count_entry_path(syscheck.database);
         w_mutex_unlock(&syscheck.fim_entry_mutex);
 
+#ifdef WIN32
+        nodes_count += fim_db_get_count_registry_key(syscheck.database) + fim_db_get_count_registry_data(syscheck.database);
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/create_db.c]: Done

File: src/syscheckd/registry/events.c [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+    cJSON *data = cJSON_CreateObject();
+    cJSON_AddItemToObject(json_event, "data", data);
+
+    snprintf(path, OS_SIZE_512, "%s\\%s", new_data->registry_entry.key->path, new_data->registry_entry.value->name);
+    cJSON_AddStringToObject(data, "path", path);
+    cJSON_AddStringToObject(data, "mode", FIM_EVENT_MODE[mode]);
+    cJSON_AddStringToObject(data, "type", FIM_EVENT_TYPE[type]);
+    cJSON_AddNumberToObject(data, "timestamp", new_data->registry_entry.value->last_event);
+
+    cJSON_AddItemToObject(data, "attributes", fim_registry_value_attributes_json(new_data->registry_entry.value, configuration));
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/events.c]: Done

File: src/syscheckd/registry/events.c [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+    }
+
+    if (*data->checksum) {
+        cJSON_AddStringToObject(attributes, "checksum", data->checksum);
+    }
+
+    return attributes;
+}
+
+cJSON *fim_registry_compare_key_attrs(const fim_registry_key *new_data, const fim_registry_key *old_data, const registry *configuration) {
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/events.c]: Done

File: src/syscheckd/registry/events.c [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+    cJSON *json_event = cJSON_CreateObject();
+    cJSON_AddStringToObject(json_event, "type", "event");
+
+    cJSON *data = cJSON_CreateObject();
+    cJSON_AddItemToObject(json_event, "data", data);
+
+    cJSON_AddStringToObject(data, "path", new_data->path);
+    cJSON_AddStringToObject(data, "mode", FIM_EVENT_MODE[mode]);
+    cJSON_AddStringToObject(data, "type", FIM_EVENT_TYPE[type]);
+    // cJSON_AddNumberToObject(data, "timestamp", new_data->last_event);
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/events.c]: Done

File: src/syscheckd/registry/events.c [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+    }
+
+    if (new->type != FIM_TYPE_REGISTRY) {
+        // This is just silly now
+        merror("LOGIC ERROR - New entry type is not Registry");
+        return NULL;
+    }
+
+    if (saved && saved->type != FIM_TYPE_REGISTRY) {
+        // This is silly too
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/events.c]: Done

File: src/syscheckd/registry/registry.c [hunk 9]
```diff
@@ -0,0 +1,501 @@
...
+            os_free(json_formated);
+
+            cJSON_Delete(json_event);
+        }
+    }
+
+    fim_db_remove_registry_value_data(fim_sql, data->registry_entry.value);
+}
+
+void fim_registry_process_key_delete_event(fdb_t *fim_sql, fim_entry *data, pthread_mutex_t *mutex, void *_alert, void *_ev_mode, void *_w_evt) {
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/registry.c]: Done

File: src/syscheckd/registry/registry.c [hunk 10]
```diff
@@ -0,0 +1,501 @@
...
+}
+
+void fim_registry_process_unscanned_entries() {
+    fim_tmp_file *file;
+    fim_event_mode event_mode = FIM_SCHEDULED;
+
+    if (fim_db_get_registry_keys_not_scanned(syscheck.database, &file, FIM_DB_DISK) == FIMDB_OK) {
+        if (file && file->elements) {
+            fim_db_process_read_file(syscheck.database, file, FIM_TYPE_REGISTRY, &syscheck.fim_registry_mutex,
+                                     fim_registry_process_key_delete_event, FIM_DB_DISK, &_base_line, &event_mode, NULL);
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/registry.c]: Done

File: src/syscheckd/registry/registry.c [hunk 11]
```diff
@@ -0,0 +1,501 @@
...
+            fim_db_process_read_registry_data_file(syscheck.database, file, &syscheck.fim_registry_mutex,
+                                                   fim_registry_process_value_delete_event, FIM_DB_DISK, &_base_line,
+                                                   &event_mode, NULL);
+        }
+    } else {
+        mwarn("Failed to get unscanned registry values");
+    }
+}
+
+void fim_registry_process_value_event(fim_entry *new, fim_entry *saved, int arch, fim_event_mode mode, BYTE *data_buffer) {
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/registry.c]: Done

File: src/syscheckd/registry/registry.c [hunk 12]
```diff
@@ -0,0 +1,501 @@
...
+        cJSON_Delete(json_event);
+    }
+    fim_db_set_registry_data_scanned(syscheck.database, new->registry_entry.value->name, new->registry_entry.key->id);
+
+    fim_registry_free_value_data(saved->registry_entry.value);
+    os_free(diff);
+}
+
+/* Query the key and get all its values */
+void fim_read_values(HKEY key_handle, fim_entry *new, fim_entry *saved, int arch, DWORD value_count, fim_event_mode mode) {
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/registry.c]: Done

File: src/syscheckd/registry/registry.c [hunk 13]
```diff
@@ -0,0 +1,501 @@
...
+    }
+
+    value_data.id = new->registry_entry.key->id;
+    new->registry_entry.value = &value_data;
+
+    for (i = 0; i < value_count; i++) {
+        value_size = MAX_VALUE_NAME;
+        data_size = MAX_VALUE_NAME;
+
+        if (RegEnumValue(key_handle, i, value_buffer, &value_size, NULL, &data_type, data_buffer, &data_size) != ERROR_SUCCESS) {
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/registry.c]: Done

File: src/syscheckd/registry/registry.c [hunk 14]
```diff
@@ -0,0 +1,501 @@
...
+    }
+
+    /* Query each sub_key and call open_key */
+    for (i = 0; i < sub_key_count; i++) {
+        char new_full_key[MAX_KEY + 2];
+        char *new_sub_key;
+        TCHAR sub_key_name_b[MAX_KEY_LENGTH + 1];
+        DWORD sub_key_name_s = MAX_KEY_LENGTH;
+
+        if (RegEnumKeyEx(current_key_handle, i, sub_key_name_b, &sub_key_name_s, NULL, NULL, NULL, NULL) != ERROR_SUCCESS) {
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/registry.c]: Done

File: src/syscheckd/registry/registry.c [hunk 15]
```diff
@@ -0,0 +1,501 @@
...
+    saved.registry_entry.key = fim_db_get_registry_key(syscheck.database, full_key);
+    saved.registry_entry.value = NULL;
+
+    if (saved.registry_entry.key != NULL) {
+        new.registry_entry.key->id = saved.registry_entry.key->id;
+    }
+
+    if (!fim_check_restrict(full_key, configuration->filerestrict)) {
+        cJSON *json_event = fim_registry_event(&new, &saved, configuration, mode,
+                                               saved.registry_entry.key == NULL ? FIM_ADD : FIM_MODIFICATION, NULL, NULL);
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/registry.c]: Done

File: src/syscheckd/registry/registry.c [hunk 16]
```diff
@@ -0,0 +1,501 @@
...
+
+        if (json_event) {
+            if (fim_db_insert_registry_key(syscheck.database, new.registry_entry.key, new.registry_entry.key->id) != FIMDB_OK) {
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/registry.c]: Done

File: src/syscheckd/registry/registry.c [hunk 17]
```diff
@@ -0,0 +1,501 @@
...
+    if (configuration->opts | CHECK_SEECHANGES) {
+        diff = fim_registry_value_diff(new->registry_entry.key->path, new->registry_entry.value->name, data_buffer,
+                                       new->registry_entry.value->type);
+    }
+
+    json_event = fim_registry_event(new, saved, configuration, mode,
+                                    saved->registry_entry.value == NULL ? FIM_ADD : FIM_MODIFIED, NULL, diff);
+
+    if (json_event) {
+        if (fim_db_insert_registry_data(syscheck.database, new->registry_entry.value, new->registry_entry.key->id) != FIMDB_OK) {
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/registry.c]: Done

File: src/syscheckd/registry/registry.c [hunk 18]
```diff
@@ -0,0 +1,501 @@
...
+        if (*syscheck.registry[i].entry == '\0') {
+            continue;
+        }
+
+        /* Read syscheck registry entry */
+        mdebug2(FIM_READING_REGISTRY, syscheck.registry[i].arch == ARCH_64BIT ? "[x64] " : "[x32] ",
+                syscheck.registry[i].entry);
+
+        if (fim_set_root_key(&root_key_handle, syscheck.registry[i].entry, &sub_key) != 0) {
+            mdebug1(FIM_INV_REG, syscheck.registry[i].entry, syscheck.registry[i].arch == ARCH_64BIT ? "[x64] " : "[x32]");
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/registry.c]: Done

File: src/syscheckd/registry/events.c [hunk 19]
```diff
@@ -0,0 +1,284 @@
...
+                          const registry *configuration,
+                          fim_event_mode mode,
+                          unsigned int event_type,
+                          __attribute__((unused)) whodata_evt *w_evt,
+                          const char *diff) {
+    cJSON *json_event = NULL;
+
+    if (new == NULL) {
+        // This should never happen
+        merror("LOGIC ERROR - new '%p' - saved '%p'", new, saved);
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/events.c]: This error message was replaced completely.

File: src/syscheckd/registry/registry.c [hunk 20]
```diff
@@ -0,0 +1,501 @@
...
+    fim_registry_value_data value_data;
+    TCHAR value_buffer[MAX_VALUE_NAME + 1];
+    DWORD value_size;
+    BYTE data_buffer[MAX_VALUE_NAME + 1];
+    DWORD data_size;
+    DWORD data_type = 0;
+    DWORD i;
+
+    if (new->registry_entry.key->id == 0) {
+        if (fim_db_get_registry_key_rowid(syscheck.database, new->registry_entry.key->path, &new->registry_entry.key->id) != FIMDB_OK) {
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/registry.c]: Done

File: src/syscheckd/syscheck.h [hunk 21]
```diff
@@ -319,7 +309,7 @@ void free_inode_data(fim_inode_data **data);
  * @brief Check the registries in the configuration for changes
  *
  */
-void os_winreg_check(void);
+void fim_registry_scan(void);
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/syscheck.h]: This prototype declaration was removed in favour of the one in `registry.h`

File: src/syscheckd/registry/registry.c [hunk 22]
```diff
@@ -0,0 +1,501 @@
...
+
+    fim_registry_process_unscanned_entries();
+
+    mdebug1(FIM_WINREGISTRY_ENDED);
+
+    if (_base_line == 0) {
+        _base_line = 1;
+    }
+
+    return;
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/registry/registry.c]: Done

File: src/syscheckd/registry/registry.c (hunk 1 above)
Comment (Molter73) [reply src/syscheckd/registry/registry.c]: Done

File: src/syscheckd/registry/events.c (hunk 2 above)
Comment (Molter73) [reply src/syscheckd/registry/events.c]: Done

File: src/syscheckd/registry/registry.c (hunk 3 above)
Comment (Molter73) [reply src/syscheckd/registry/registry.c]: Done

**Review by jesusjimsa — APPROVED** [review]
LGTM!

**Pull request merged**